File: proj/CoreLibrary/src/com/jasonzqshen/familyaccounting/core/masterdata/MasterDataManagement.java
package com.jasonzqshen.familyaccounting.core.masterdata;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.Hashtable;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;

import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;

import com.jasonzqshen.familyaccounting.core.CoreDriver;
import com.jasonzqshen.familyaccounting.core.ManagementBase;
import com.jasonzqshen.familyaccounting.core.exception.IdentityInvalidChar;
import com.jasonzqshen.familyaccounting.core.exception.IdentityNoData;
import com.jasonzqshen.familyaccounting.core.exception.IdentityTooLong;
import com.jasonzqshen.familyaccounting.core.exception.format.MasterDataFileFormatException;
import com.jasonzqshen.familyaccounting.core.exception.runtime.NoMasterDataFactoryClass;
import com.jasonzqshen.familyaccounting.core.exception.runtime.SystemException;
import com.jasonzqshen.familyaccounting.core.utils.GLAccountGroup;
import com.jasonzqshen.familyaccounting.core.utils.MessageType;

/**
 * Master Data Management, which contains all master data factories.
 * 
 * @author I072485
 * 
 */
public class MasterDataManagement extends ManagementBase {
	public static final String MASTER_DATA_FOLDER = "master_data";

	private final Hashtable<MasterDataType, MasterDataFactoryBase> _factoryList;
	private final Hashtable<MasterDataType, Class<?>> _registerFactorys;

	/**
	 * 
	 * @param coreDriver
	 */
	public MasterDataManagement(CoreDriver coreDriver) {
		super(coreDriver);
		_factoryList = new Hashtable<MasterDataType, MasterDataFactoryBase>();
		_registerFactorys = new Hashtable<MasterDataType, Class<?>>();
		// initialize the hash table
		registerFactory();

	}

	/**
	 * register master data factory
	 */
	private void registerFactory() {
		_registerFactorys.put(MasterDataType.VENDOR,
				VendorMasterDataFactory.class);
		_registerFactorys.put(MasterDataType.CUSTOMER,
				CustomerMasterDataFactory.class);
		_registerFactorys.put(MasterDataType.BANK_KEY,
				BankKeyMasterDataFactory.class);
		_registerFactorys.put(MasterDataType.BUSINESS_AREA,
				BusinessAreaMasterDataFactory.class);
		_registerFactorys.put(MasterDataType.BANK_ACCOUNT,
				BankAccountMasterDataFactory.class);
		_registerFactorys.put(MasterDataType.GL_ACCOUNT,
				GLAccountMasterDataFactory.class);
	}

	/**
	 * parse XML document to master data factory
	 * 
	 * @param constructor
	 * @param type
	 * @param doc
	 * @param messages
	 * @return
	 * @throws MasterDataFileFormatException
	 */
	private MasterDataFactoryBase factoryParser(Constructor<?> constructor,
			MasterDataType type, Document doc)
			throws MasterDataFileFormatException {
		_coreDriver.logDebugInfo(this.getClass(), 88,
				String.format("Parsing XML %s...", type), MessageType.INFO);

		// get root element
		NodeList nodeList = doc.getChildNodes();
		Element rootElem = null;

		// get root element
		for (int i = 0; i < nodeList.getLength(); ++i) {
			Node child = nodeList.item(i);
			if (child instanceof Element) {
				Element elem = (Element) child;
				String nodeName = elem.getNodeName();
				if (nodeName.equals(MasterDataUtils.XML_ROOT)) {
					rootElem = elem;
					break;
				}
			}
		}
		// no root element
		if (rootElem == null) {
			_coreDriver.logDebugInfo(this.getClass(), 112,
					String.format("No root element", type), MessageType.ERROR);
			throw new MasterDataFileFormatException(type);
		}

		MasterDataFactoryBase newFactory = null;
		try {
			newFactory = (MasterDataFactoryBase) constructor.newInstance(
					_coreDriver, this);
			nodeList = rootElem.getChildNodes();
			for (int i = 0; i < nodeList.getLength(); ++i) {
				Node child = nodeList.item(i);
				if (child instanceof Element) {
					Element elem = (Element) child;
					if (elem.getNodeName().equals(MasterDataUtils.XML_ENTITY)) {
						// parse master data entity
						MasterDataBase masterData = newFactory.parseMasterData(
								_coreDriver, elem);

						// raise load master data
						_coreDriver.getListenersManagement().loadMasterData(
								this, masterData);
					}
				}
			}
		} catch (InstantiationException e) {
			_coreDriver.logDebugInfo(this.getClass(), 133, e.toString(),
					MessageType.ERROR);
			throw new SystemException(e);
		} catch (IllegalArgumentException e) {
			_coreDriver.logDebugInfo(this.getClass(), 137, e.toString(),
					MessageType.ERROR);
			throw new SystemException(e);
		} catch (IllegalAccessException e) {
			_coreDriver.logDebugInfo(this.getClass(), 140, e.toString(),
					MessageType.ERROR);
			throw new SystemException(e);
		} catch (InvocationTargetException e) {
			_coreDriver.logDebugInfo(this.getClass(), 145, e.toString(),
					MessageType.ERROR);
			throw new SystemException(e);
		} catch (MasterDataFileFormatException e) {
			_coreDriver.logDebugInfo(this.getClass(), 112,
					String.format("Mandatory fields contains no data", type),
					MessageType.ERROR);
			throw new MasterDataFileFormatException(type);
		}

		_coreDriver.logDebugInfo(this.getClass(), 88,
				String.format("Parsing XML %s complete.", type),
				MessageType.INFO);

		return newFactory;
	}

	/**
	 * load the data from file system
	 * 
	 * @throws NoMasterDataFactoryClass
	 *             if the exception raised from the function, it is the bug.
	 * @throws SystemException
	 *             if the exception raised from the function, it is the bug.
	 * 
	 */
	public void initialize() throws MasterDataFileFormatException {
		_coreDriver.logDebugInfo(this.getClass(), 141,
				"Master data loading...", MessageType.INFO);

		for (MasterDataType type : MasterDataType.values()) {
			_coreDriver.logDebugInfo(this.getClass(), 144,
					String.format("Loading %s...", type), MessageType.INFO);

			String filePath = generateMasterFilePath(type);
			_coreDriver.logDebugInfo(this.getClass(), 148, String.format(
					"Master data file %s for master data %s.", filePath, type),
					MessageType.INFO);

			// get factory class
			Class<?> factoryClass = _registerFactorys.get(type);
			if (factoryClass == null) {
				_coreDriver
						.logDebugInfo(
								this.getClass(),
								178,
								String.format(
										"Master data class is not registered for master data %s",
										type), MessageType.WARNING);
				throw new NoMasterDataFactoryClass(type);
			}
			// get constructor
			Constructor<?> constructor = null;
			try {
				constructor = factoryClass.getConstructor(CoreDriver.class,
						MasterDataManagement.class);
			} catch (NoSuchMethodException e1) {
				_coreDriver.logDebugInfo(this.getClass(), 191, e1.toString(),
						MessageType.ERROR);
				throw new SystemException(e1);
			} catch (SecurityException e1) {
				_coreDriver.logDebugInfo(this.getClass(), 193, e1.toString(),
						MessageType.ERROR);
				throw new SystemException(e1);
			}

			// open file
			File file = new File(filePath);
			if (!file.exists()) {
				_coreDriver
						.logDebugInfo(
								this.getClass(),
								165,
								String.format(
										"Master data file %s for master data %s does not exist.",
										filePath, type), MessageType.ERROR);
				throw new MasterDataFileFormatException(type);
			}

			try {
				DocumentBuilderFactory docFactory = DocumentBuilderFactory
						.newInstance();
				DocumentBuilder builder = docFactory.newDocumentBuilder();
				Document doc = builder.parse(file);

				// parse
				MasterDataFactoryBase newFactory = factoryParser(constructor,
						type, doc);

				_coreDriver.logDebugInfo(this.getClass(), 208,
						String.format("Loading %s completed.", type),
						MessageType.INFO);

				this._factoryList.put(type, newFactory);
			} catch (ParserConfigurationException e) {
				_coreDriver.logDebugInfo(this.getClass(), 207, e.toString(),
						MessageType.ERROR);
				throw new SystemException(e);
			} catch (SAXException e) {
				_coreDriver.logDebugInfo(this.getClass(), 216, String.format(
						"Master data file %s contains format error", filePath),
						MessageType.ERROR);
				throw new MasterDataFileFormatException(type);
			} catch (IOException e) {
				_coreDriver.logDebugInfo(this.getClass(), 220,
						String.format(e.toString(), filePath),
						MessageType.ERROR);
				throw new SystemException(e);
			}

		}

		_coreDriver.logDebugInfo(
				this.getClass(),
				282,
				"Master data loading completed. Items: "
						+ _registerFactorys.size(), MessageType.INFO);
	}

	/**
	 * create instance via the constructor
	 * 
	 * @param constructor
	 * @return
	 * @throws SystemException
	 *             system bug
	 */
	private MasterDataFactoryBase createFactory(Constructor<?> constructor) {
		try {
			MasterDataFactoryBase newFactory = (MasterDataFactoryBase) constructor
					.newInstance(_coreDriver, this);

			return newFactory;
		} catch (IllegalArgumentException e) {
			_coreDriver.logDebugInfo(this.getClass(), 240, e.toString(),
					MessageType.ERROR);
			throw new SystemException(e);
		} catch (InstantiationException e) {
			_coreDriver.logDebugInfo(this.getClass(), 244, e.toString(),
					MessageType.ERROR);
			throw new SystemException(e);
		} catch (IllegalAccessException e) {
			_coreDriver.logDebugInfo(this.getClass(), 248, e.toString(),
					MessageType.ERROR);
			throw new SystemException(e);
		} catch (InvocationTargetException e) {
			_coreDriver.logDebugInfo(this.getClass(), 251, e.toString(),
					MessageType.ERROR);
			throw new SystemException(e);
		}
	}

	/**
	 * store master data
	 * 
	 * @throws SystemException
	 */
	public void store() throws SystemException {
		for (MasterDataType type : MasterDataType.values()) {
			storeSingle(type);
		}
	}

	/**
	 * store one type master data
	 * 
	 * @param type
	 * @throws SystemException
	 */
	public void storeSingle(MasterDataType type) {
		_coreDriver.logDebugInfo(this.getClass(), 335,
				String.format("Starting to store master data %s...", type),
				MessageType.INFO);

		MasterDataFactoryBase factory = _factoryList.get(type);

		if (factory._containDirtyData == false) {
			_coreDriver
					.logDebugInfo(
							this.getClass(),
							335,
							"No dirty data in the memory. No need to save the data to disk.",
							MessageType.INFO);
			return;
		}
		// get XML document
		String xdoc = factory.toXmlDocument();

		File file = new File(generateMasterFilePath(type));
		if (!file.exists()) {
			try {
				file.createNewFile();
			} catch (IOException e) {
				_coreDriver.logDebugInfo(this.getClass(), 335, e.toString(),
						MessageType.ERROR);
				throw new SystemException(e);
			}
		}

		FileWriter writer;
		try {
			writer = new FileWriter(file);
			writer.write(xdoc, 0, xdoc.length());
			writer.close();
		} catch (IOException e) {
			_coreDriver.logDebugInfo(this.getClass(), 335, e.toString(),
					MessageType.ERROR);
			throw new SystemException(e);
		}

		factory._containDirtyData = false;
		_coreDriver.logDebugInfo(this.getClass(), 335,
				String.format("Store master data %s complete.", type),
				MessageType.INFO);
	}

	/**
	 * get path
	 * 
	 * @param path
	 * @return
	 */
	private String generateMasterFilePath(MasterDataType type) {
		String path = null;
		if (type == MasterDataType.VENDOR) {
			path = VendorMasterData.FILE_NAME;
		} else if (type == MasterDataType.CUSTOMER) {
			path = CustomerMasterData.FILE_NAME;
		} else if (type == MasterDataType.BUSINESS_AREA) {
			path = BusinessAreaMasterData.FILE_NAME;
		} else if (type == MasterDataType.BANK_KEY) {
			path = BankKeyMasterData.FILE_NAME;
		} else if (type == MasterDataType.BANK_ACCOUNT) {
			path = BankAccountMasterData.FILE_NAME;
		} else if (type == MasterDataType.GL_ACCOUNT) {
			path = GLAccountMasterData.FILE_NAME;
		}

		return String.format("%s/%s/%s", _coreDriver.getRootPath(),
				MASTER_DATA_FOLDER, path);
	}

	/**
	 * get master data factory
	 * 
	 * @param type
	 *            master data type
	 * @return master data factory
	 */
	public MasterDataFactoryBase getMasterDataFactory(MasterDataType type) {
		return this._factoryList.get(type);
	}

	/**
	 * get the entity of master data
	 * 
	 * @param idStr
	 *            identity of master data
	 * @param type
	 *            type of master data
	 * @return master data entity
	 * @throws IdentityInvalidChar
	 * @throws IdentityNoData
	 * @throws IdentityTooLong
	 */
	public MasterDataBase getMasterData(char[] idStr, MasterDataType type)
			throws IdentityTooLong, IdentityNoData, IdentityInvalidChar {
		MasterDataFactoryBase factory = this.getMasterDataFactory(type);
		MasterDataIdentity id = new MasterDataIdentity(idStr);

		return factory.getEntity(id);
	}

	/**
	 * get master data based on identity and data type
	 * 
	 * @param id
	 *            identity
	 * @param type
	 *            master data type
	 * @return
	 * 
	 */
	public MasterDataBase getMasterData(MasterDataIdentity id,
			MasterDataType type) {
		MasterDataFactoryBase factory = this.getMasterDataFactory(type);
		return factory.getEntity(id);
	}

	/**
	 * contains master data
	 * 
	 * @param id
	 * @param type
	 * @return
	 */
	public boolean containsMasterData(MasterDataIdentity id, MasterDataType type) {
		MasterDataFactoryBase factory = this.getMasterDataFactory(type);
		return factory.contains(id);
	}

	/**
	 * get G/L accounts based on G/L account group
	 * 
	 * @param group
	 *            G/L account group
	 * @return
	 */
	public MasterDataIdentity_GLAccount[] getGLAccountsBasedGroup(
			GLAccountGroup group) {
		MasterDataFactoryBase factory = getMasterDataFactory(MasterDataType.GL_ACCOUNT);
		MasterDataBase[] datas = factory.getAllEntities();
		ArrayList<MasterDataIdentity_GLAccount> array = new ArrayList<MasterDataIdentity_GLAccount>();

		for (MasterDataBase data : datas) {
			GLAccountMasterData glAccount = (GLAccountMasterData) data;
			if (glAccount.getGroup() == group) {
				array.add(glAccount.getGLIdentity());
			}
		}

		MasterDataIdentity_GLAccount[] ret = new MasterDataIdentity_GLAccount[array
				.size()];
		for (int i = 0; i < ret.length; ++i) {
			ret[i] = array.get(i);
		}
		return ret;
	}

	/**
	 * get G/L accounts based on G/L account group
	 * 
	 * @param group
	 *            G/L account group
	 * @return
	 */
	public GLAccountMasterData[] getBalanceAccounts() {
		GLAccountMasterDataFactory factory = (GLAccountMasterDataFactory) getMasterDataFactory(MasterDataType.GL_ACCOUNT);
		return factory.getBalanceAccounts();
	}

	/**
	 * get liquidity
	 * 
	 * @param group
	 * @return
	 */
	public GLAccountMasterData[] getLiquidityAccounts() {
		GLAccountMasterDataFactory factory = (GLAccountMasterDataFactory) getMasterDataFactory(MasterDataType.GL_ACCOUNT);
		return factory.getLiquidityAccounts();
	}

	/**
	 * get liability
	 * 
	 * @param group
	 * @return
	 */
	public GLAccountMasterData[] getLiabilityAccounts() {
		GLAccountMasterDataFactory factory = (GLAccountMasterDataFactory) getMasterDataFactory(MasterDataType.GL_ACCOUNT);
		return factory.getLiabilityAccounts();
	}

	/**
	 * get cost accounts
	 * 
	 * @param group
	 * @return
	 */
	public GLAccountMasterData[] getCostAccounts() {
		GLAccountMasterDataFactory factory = (GLAccountMasterDataFactory) getMasterDataFactory(MasterDataType.GL_ACCOUNT);
		return factory.getCostAccounts();
	}

	/**
	 * get revenue accounts
	 * 
	 * @param group
	 * @return
	 */
	public GLAccountMasterData[] getRevenueAccounts() {
		GLAccountMasterDataFactory factory = (GLAccountMasterDataFactory) getMasterDataFactory(MasterDataType.GL_ACCOUNT);
		return factory.getRevenueAccounts();
	}

	/**
	 * clear
	 */
	public void clear() {
		_factoryList.clear();
	}

	@Override
	public void establishFiles() {
		String masterFolderPath = String.format("%s/%s",
				_coreDriver.getRootPath(), MASTER_DATA_FOLDER);

		// set up master data folder
		File masterFolder = new File(masterFolderPath);
		if (!masterFolder.exists()) {
			_coreDriver.logDebugInfo(this.getClass(), 113,
					"Master data root folder does not exist. Make directory.",
					MessageType.INFO);
			masterFolder.mkdir();
		}

		// establish master data files
		for (MasterDataType type : MasterDataType.values()) {
			// get factory class
			Class<?> factoryClass = _registerFactorys.get(type);
			if (factoryClass == null) {
				_coreDriver
						.logDebugInfo(
								this.getClass(),
								178,
								String.format(
										"Master data class is not registered for master data %s",
										type), MessageType.ERROR);
				throw new NoMasterDataFactoryClass(type);
			}

			try {
				Constructor<?> constructor = factoryClass.getConstructor(
						CoreDriver.class, MasterDataManagement.class);
				// create new factory
				MasterDataFactoryBase factory = createFactory(constructor);
				_factoryList.put(type, factory);

				String xdoc = factory.toXmlDocument();

				String filePath = generateMasterFilePath(type);

				File file = new File(filePath);
				if (!file.exists()) {
					file.createNewFile();
				}

				FileWriter writer = new FileWriter(file);
				writer.write(xdoc, 0, xdoc.length());
				writer.close();
			} catch (NoSuchMethodException e1) {
				_coreDriver.logDebugInfo(this.getClass(), 191, e1.toString(),
						MessageType.ERROR);
				throw new SystemException(e1);
			} catch (SecurityException e1) {
				_coreDriver.logDebugInfo(this.getClass(), 193, e1.toString(),
						MessageType.ERROR);
				throw new SystemException(e1);
			} catch (IOException e) {
				_coreDriver.logDebugInfo(this.getClass(), 193, e.toString(),
						MessageType.ERROR);
				throw new SystemException(e);
			}
		}

	}
}

File: proj/WinStore_FamilyAccounting/WinStore_FamilyAccountingCore/MasterData/MasterDataManagement.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Security;
using System.Text;
using System.Threading.Tasks;
using System.Xml;
using System.Xml.Linq;
using Windows.Storage;
using Windows.Storage.Streams;
using WinStore_FamilyAccountingCore.Exceptions;
using WinStore_FamilyAccountingCore.Exceptions.FormatExceptions;
using WinStore_FamilyAccountingCore.Utilities;

namespace WinStore_FamilyAccountingCore.MasterData
{
    /// <summary>
    /// master data utilities
    /// </summary>
    public class MasterDataUtils
    {
        public static readonly String DATE_FORMAT = "yyyy.MM.dd";

        public static readonly String XML_ID = "id"; // identity
        public static readonly String XML_DESCP = "name"; // description
        public static readonly String XML_ROOT = "root";
        public static readonly String XML_ENTITY = "entity";
        public static readonly String XML_CRITICAL_LEVEL = "critical"; // critical
        // level
        public static readonly String XML_BANK_KEY = "bank_key"; // bank key
        public static readonly String XML_BANK_ACCOUNT = "bank_account"; // bank
        public static readonly String XML_INIT_AMOUNT = "init_amount"; // init amount
        // account
        public static readonly String XML_TYPE = "type";


        private MasterDataUtils()
        {
        }
    }

    /// <summary>
    /// master data type
    /// </summary>
    public enum MasterDataType
    {
        VENDOR, CUSTOMER, BUSINESS_AREA, BANK_KEY, BANK_ACCOUNT, GL_ACCOUNT
    }

    /// <summary>
    /// master data managment
    /// </summary>
    public class MasterDataManagement : AbstractManagement
    {
        public static readonly String MASTER_DATA_FOLDER = "master_data";

        private readonly Dictionary<MasterDataType, MasterDataFactoryBase> _factoryList;

        /// <summary>
        /// balance accounts
        /// </summary>
        public List<GLAccountMasterData> BalanceAccounts
        {
            get
            {
                GLAccountMasterDataFactory factory = (GLAccountMasterDataFactory)
                    GetMasterDataFactory(MasterDataType.GL_ACCOUNT);
                return factory.BalanceAccounts;
            }
        }

        /// <summary>
        /// Liquidity accounts
        /// </summary>
        public List<GLAccountMasterData> LiquidityAccounts
        {
            get
            {
                GLAccountMasterDataFactory factory = (GLAccountMasterDataFactory)
                    GetMasterDataFactory(MasterDataType.GL_ACCOUNT);
                return factory.LiabilityAccounts;
            }
        }

        /// <summary>
        /// Liability accounts
        /// </summary>
        public List<GLAccountMasterData> LiabilityAccounts
        {
            get
            {
                GLAccountMasterDataFactory factory = (GLAccountMasterDataFactory)
                    GetMasterDataFactory(MasterDataType.GL_ACCOUNT);
                return factory.LiabilityAccounts;
            }
        }

        /// <summary>
        /// Cost accounts
        /// </summary>
        public List<GLAccountMasterData> CostAccounts
        {
            get
            {
                GLAccountMasterDataFactory factory = (GLAccountMasterDataFactory)
                    GetMasterDataFactory(MasterDataType.GL_ACCOUNT);
                return factory.CostAccounts;
            }
        }

        /// <summary>
        /// Revenue accounts
        /// </summary>
        public List<GLAccountMasterData> RevenueAccounts
        {
            get
            {
                GLAccountMasterDataFactory factory = (GLAccountMasterDataFactory)
                    GetMasterDataFactory(MasterDataType.GL_ACCOUNT);
                return factory.RevenueAccounts;
            }
        }

        /// <summary>
        /// Master data management
        /// </summary>
        /// <param name="coreDriver"></param>
        internal MasterDataManagement(CoreDriver coreDriver)
            : base(coreDriver)
        {
            _factoryList = new Dictionary<MasterDataType, MasterDataFactoryBase>();
            // initialize the hash table
        }


        /// <summary>
        /// parse XDocument to master data factory
        /// </summary>
        /// <param name="constructor"></param>
        /// <param name="type"></param>
        /// <param name="doc"></param>
        /// <returns></returns>
        /// <exception cref="MasterDataFileFormatException">Master data file format Exception</exception>
        /// <exception cref="SystemException">Bug</exception>
        private MasterDataFactoryBase factoryParser(
            MasterDataType type, XDocument doc)
        {
            _coreDriver.logDebugInfo(this.GetType(), 88,
                    String.Format("Parsing XML {0}...", type), MessageType.INFO);

            // get root element
            XElement rootElem = doc.Element(MasterDataUtils.XML_ROOT);
            if (rootElem == null)
            {
                _coreDriver.logDebugInfo(this.GetType(), 112,
                        String.Format("No root element for master data {0}", type), MessageType.ERRO);
                throw new MasterDataFileFormatException(type);
            }

            MasterDataFactoryBase newFactory;
            #region Create new factory with contructor
            switch (type)
            {
                case MasterDataType.BANK_ACCOUNT:
                    newFactory = new BankAccountMasterDataFactory(_coreDriver, this);
                    break;
                case MasterDataType.BANK_KEY:
                    newFactory = new BankKeyMasterDataFactory(_coreDriver, this);
                    break;
                case MasterDataType.BUSINESS_AREA:
                    newFactory = new BusinessAreaMasterDataFactory(_coreDriver, this);
                    break;
                case MasterDataType.CUSTOMER:
                    newFactory = new CustomerMasterDataFactory(_coreDriver, this);
                    break;
                case MasterDataType.GL_ACCOUNT:
                    newFactory = new GLAccountMasterDataFactory(_coreDriver, this);
                    break;
                case MasterDataType.VENDOR:
                    newFactory = new VendorMasterDataFactory(_coreDriver, this);
                    break;
                default:
                    throw new SystemException(new NoMasterDataFactoryClass(type));
            }

            #endregion

            try
            {
                foreach (XElement xelem in rootElem.Elements())
                {
                    if (xelem.Name.LocalName.Equals(MasterDataUtils.XML_ENTITY))
                    {
                        // parse master data entity
                        MasterDataBase masterData = newFactory.ParseMasterData(
                                _coreDriver, xelem);

                        // raise load master data
                        _coreDriver.ListenerMgmt.LoadMasterData(
                                this, masterData);
                    }
                }
            }
            catch (ArgumentNullException e)
            {
                _coreDriver.logDebugInfo(this.GetType(), 133, e.Message,
                        MessageType.ERRO);
                throw new SystemException(e);
            }

            _coreDriver.logDebugInfo(this.GetType(), 88,
                    String.Format("Parsing XML {0} complete.", type),
                    MessageType.INFO);
            return newFactory;
        }

        /// <summary>
        /// load the data from file system
        /// </summary>
        /// <returns></returns>
        /// <exception cref="MasterDataFileFormatException"></exception>
        /// <exception cref="SystemException"></exception>
        public override async Task InitializeAsync()
        {
            _coreDriver.logDebugInfo(this.GetType(), 141,
                    "Master data loading...", MessageType.INFO);
            foreach (MasterDataType type in Enum.GetValues(typeof(MasterDataType)))
            {
                _coreDriver.logDebugInfo(this.GetType(), 144,
                        String.Format("Loading {0}...", type), MessageType.INFO);

                StorageFile file = await getMasterFileAsync(type);
                _coreDriver.logDebugInfo(this.GetType(), 148, String.Format(
                        "Master data file {0} for master data {1}.", file.Path, type),
                        MessageType.INFO);

                String str = await FileIO.ReadTextAsync(file);
                XDocument xdoc = XDocument.Parse(str);

                // parse
                MasterDataFactoryBase newFactory = factoryParser(type, xdoc);

                _coreDriver.logDebugInfo(this.GetType(), 208,
                        String.Format("Loading %s completed.", type),
                        MessageType.INFO);

                this._factoryList.Add(type, newFactory);
            }

            _coreDriver.logDebugInfo(
                    this.GetType(),
                    282,
                    "Master data loading completed.", MessageType.INFO);
        }

        /// <summary>
        /// Get master data factory
        /// </summary>
        /// <param name="type"></param>
        /// <returns></returns>
        public MasterDataFactoryBase GetMasterDataFactory(MasterDataType type)
        {
            if (!this._factoryList.ContainsKey(type))
            {
                return null;
            }
            MasterDataFactoryBase factory;
            this._factoryList.TryGetValue(type, out factory);
            return factory;
        }

        /// <summary>
        /// get master data with identity
        /// </summary>
        /// <param name="idStr"></param>
        /// <param name="type"></param>
        /// <returns></returns>
        /// <exception cref="IdentityInvalidChar"></exception>
        /// <exception cref="IdentityNoData"></exception>
        /// <exception cref="IdentityTooLong"></exception>
        public MasterDataBase GetMasterData(char[] idStr, MasterDataType type)
        {
            MasterDataFactoryBase factory = this.GetMasterDataFactory(type);
            MasterDataIdentity id = new MasterDataIdentity(idStr);

            return factory.GetEntity(id);
        }

        /// <summary>
        /// get master data 
        /// </summary>
        /// <param name="id"></param>
        /// <param name="type"></param>
        /// <returns></returns>
        public MasterDataBase GetMasterData(MasterDataIdentity id,
                MasterDataType type)
        {
            MasterDataFactoryBase factory = this.GetMasterDataFactory(type);
            return factory.GetEntity(id);
        }

        /// <summary>
        /// constain master data
        /// </summary>
        /// <param name="id"></param>
        /// <param name="type"></param>
        /// <returns></returns>
        public bool ContainsMasterData(MasterDataIdentity id, MasterDataType type)
        {
            MasterDataFactoryBase factory = this.GetMasterDataFactory(type);
            return factory.Contains(id);
        }

        /// <summary>
        /// Get GL Account based on G/L account group
        /// </summary>
        /// <param name="group"></param>
        /// <returns></returns>
        public List<MasterDataIdentity_GLAccount> GetGLAccountsBasedGroup(
                GLAccountGroupENUM group)
        {
            MasterDataFactoryBase factory = GetMasterDataFactory(MasterDataType.GL_ACCOUNT);
            List<MasterDataBase> datas = factory.AllEntities;
            List<MasterDataIdentity_GLAccount> array = new List<MasterDataIdentity_GLAccount>();

            foreach (MasterDataBase data in datas)
            {
                GLAccountMasterData glAccount = (GLAccountMasterData)data;
                if (glAccount.Group.Identity == group)
                {
                    array.Add(glAccount.GLIdentity);
                }
            }

            array.Sort();
            return array;
        }

        /// <summary>
        /// clear memory
        /// </summary>
        public override void Clear()
        {
            _factoryList.Clear();
        }

        /// <summary>
        /// establish master data files
        /// </summary>
        /// <returns></returns>
        public override async Task EstablishFilesAsync()
        {
            StorageFolder masterFolder = await _coreDriver.RootFolder.CreateFolderAsync(MASTER_DATA_FOLDER
                , CreationCollisionOption.OpenIfExists);

            // establish master data files
            foreach (MasterDataType type in Enum.GetValues(typeof(MasterDataType)))
            {
                MasterDataFactoryBase newFactory;
                #region Create new factory with contructor
                switch (type)
                {
                    case MasterDataType.BANK_ACCOUNT:
                        newFactory = new BankAccountMasterDataFactory(_coreDriver, this);
                        break;
                    case MasterDataType.BANK_KEY:
                        newFactory = new BankKeyMasterDataFactory(_coreDriver, this);
                        break;
                    case MasterDataType.BUSINESS_AREA:
                        newFactory = new BusinessAreaMasterDataFactory(_coreDriver, this);
                        break;
                    case MasterDataType.CUSTOMER:
                        newFactory = new CustomerMasterDataFactory(_coreDriver, this);
                        break;
                    case MasterDataType.GL_ACCOUNT:
                        newFactory = new GLAccountMasterDataFactory(_coreDriver, this);
                        break;
                    case MasterDataType.VENDOR:
                        newFactory = new VendorMasterDataFactory(_coreDriver, this);
                        break;
                    default:
                        throw new SystemException(new NoMasterDataFactoryClass(type));
                }
                _factoryList.Add(type, newFactory);
                #endregion

                XDocument xdoc = newFactory.ToXmlDocument();
                StorageFile file = await getMasterFileAsync(type);
                await FileIO.WriteTextAsync(file, xdoc.ToString());
            }
        }

        /// <summary>
        /// get master data file path
        /// </summary>
        /// <param name="type"></param>
        /// <returns></returns>
        private async Task<StorageFile> getMasterFileAsync(MasterDataType type)
        {
            String path = null;
            if (type == MasterDataType.VENDOR)
            {
                path = VendorMasterData.FILE_NAME;
            }
            else if (type == MasterDataType.CUSTOMER)
            {
                path = CustomerMasterData.FILE_NAME;
            }
            else if (type == MasterDataType.BUSINESS_AREA)
            {
                path = BusinessAreaMasterData.FILE_NAME;
            }
            else if (type == MasterDataType.BANK_KEY)
            {
                path = BankKeyMasterData.FILE_NAME;
            }
            else if (type == MasterDataType.BANK_ACCOUNT)
            {
                path = BankAccountMasterData.FILE_NAME;
            }
            else if (type == MasterDataType.GL_ACCOUNT)
            {
                path = GLAccountMasterData.FILE_NAME;
            }

            StorageFolder folder = await _coreDriver.RootFolder
                .GetFolderAsync(MASTER_DATA_FOLDER);
            if (folder == null)
            {
                return null;
            }
            return await folder.CreateFileAsync(path, CreationCollisionOption.OpenIfExists);
        }

        /// <summary>
        /// Store
        /// </summary>
        public async Task StoreAsync()
        {
            foreach (MasterDataType type in Enum.GetValues(typeof(MasterDataType)))
            {
                await StoreSingleAsync(type);
            }
        }

        /// <summary>
        /// store single master data
        /// </summary>
        /// <param name="type"></param>
        /// <exception cref="SystemException">Bug</exception>
        public async Task StoreSingleAsync(MasterDataType type)
        {
            _coreDriver.logDebugInfo(this.GetType(), 335,
                    String.Format("Starting to store master data {0}...", type),
                    MessageType.INFO);

            MasterDataFactoryBase factory;
            if (!_factoryList.TryGetValue(type, out factory))
            {
                throw new SystemException(new MasterDataFileFormatException(type));
            }

            if (factory._containDirtyData == false)
            {
                _coreDriver
                        .logDebugInfo(
                                this.GetType(),
                                335,
                                "No dirty data in the memory. No need to save the data to disk.",
                                MessageType.INFO);
                return;
            }
            // get XML document
            XDocument xdoc = factory.ToXmlDocument();
            StorageFile file = await this.getMasterFileAsync(type);
            try
            {
                await FileIO.WriteTextAsync(file, xdoc.ToString());
            }
            catch (FileNotFoundException e)
            {
                throw new SystemException(e);
            }

            factory._containDirtyData = false;
            _coreDriver.logDebugInfo(this.GetType(), 335,
                    String.Format("Store master data %s complete.", type),
                    MessageType.INFO);
        }
    }
}